### PR TITLE
First cut of non-recursive rules

### PR DESCRIPTION
**Pull request by FiV0**
This is an first attempt at non-recursive rules. 

The approach is first doing some variable replacement in the rule definitions. This is to assure we are not shadowing any variables from the main query. After that rules get sort of `macroexpanded` into `union-joins`. The "annoying" bit is pushing down the literals as these need to be removed as arguments from the different query operators (`exists?`, `not-exists?` `union-join` etc).

There is still some work to do on this. Mainly
- Fix https://github.com/xtdb/core2/issues/638, because this is currently blocking the proper testing of subqueies in rules.
- There is also some missing rewriting bits for subqueries (`order-by`, `aggregates`, `keys` etc...).

Concerning the semantics of rules in Core1 I am just linking my notes from the other day.
[rules_scratch.md](https://github.com/xtdb/core2/files/10827322/rules_scratch.md)

### CODE REVIEW

File: core/src/main/clojure/core2/core/datalog.clj [hunk 1]
```diff
@@ -30,7 +29,7 @@
 (s/def ::form
   (s/or :logic-var ::logic-var
         :fn-call ::fn-call
-        :value ::value))
+        :literal ::value))
 
 (s/def ::find-arg
   (s/or :logic-var ::logic-var
```
Comment (FiV0) [inline core/src/main/clojure/core2/core/datalog.clj:35]: @jms was wondering if the specs above (`fing-arg` and `form`) could be collpased or if there is something I am missing about the extra indirection?

**Comment by FiV0** [comment]
@jarohen Good for review from my side. One thing I don't like about the PR is that it relies heavily on the shape of the specs. So whenever those change one needs to update this code as well.

File: src/test/clojure/core2/core/datalog_test.clj [hunk 2]
```diff
@@ -938,6 +939,267 @@
...
+                                                   (not-exists? [age]
+                                                                [(< age 21)])]])))
+
+  (let [tx (c2/submit-tx tu/*node* [[:put {:id :ivan :name "Ivan" :last-name "Ivanov" :age 21}]
+                                    [:put {:id :petr :name "Petr" :last-name "Petrov" :age 18}]
+                                    [:put {:id :georgy :name "Georgy" :last-name "George" :age 17}]])
+        q (fn q [query & args]
+            (apply c2/q tu/*node*
+                   (-> query
+                       (assoc :basis {:tx tx}))
```
Comment (jarohen) [inline src/test/clojure/core2/core/datalog_test.clj:974]: shouldn't need this `assoc :basis` now, unless you're querying at different tx-times

File: src/test/clojure/core2/core/datalog_test.clj [hunk 3]
```diff
@@ -938,6 +939,267 @@
...
+               (q '{:find [i other-age]
+                    :where [[i :age age]
+                            (older-ages age other-age)]
+                    :rules [[(older-ages age other-age)
+                             (q {:find [other-age]
+                                 :in [age]
+                                 :where [[i :age other-age]
+                                         [(> other-age age)]]})]]}))))
+
+    ;; result needs to be updated after #663 gets fixed
```
Comment (jarohen) [inline src/test/clojure/core2/core/datalog_test.clj:1110]: done?

**Review by jarohen — APPROVED** [review]
nice one @fiv0 :smile: :pray: 

few minor, mostly subjective suggestions, but overall the approach looks sound and thorough, thanks!

File: core/src/main/clojure/core2/core/datalog.clj [hunk 4]
```diff
@@ -250,9 +266,20 @@
             (case form-type
               :fn-call (into #{} (mapcat form-vars*) (:args form-arg))
               :logic-var #{form-arg}
-              :value #{}))]
+              :literal #{}
+              :form (form-vars form-arg)))]
     (form-vars* form)))
 
+(defn- binding-vars [binding]
+  (letfn [(binding-vars* [[binding-type binding-arg]]
```
[the remark this answers is not in the document]
Comment (FiV0) [reply core/src/main/clojure/core2/core/datalog.clj:274]: Mainly just copied the style from form-vars.

File: src/test/clojure/core2/core/datalog_test.clj [hunk 5]
```diff
@@ -938,6 +939,267 @@
                            (assoc :basis {:tx !tx}))))
               "b is unified")))))
 
+(deftest test-basic-rules
+  (t/is (= '[[:triple {:e [:logic-var i], :a :age, :v [:logic-var age]}]
```
[the remark this answers is not in the document]
Comment (FiV0) [reply src/test/clojure/core2/core/datalog_test.clj:943]: I just pulled them over from C1, can remove them if you want.

File: core/src/main/clojure/core2/core/datalog.clj [hunk 6]
```diff
@@ -599,10 +631,243 @@
...
     (mega-join (into [plan] sub-queries) param-vars)))
 
-(defn- plan-body [{where-clauses :where, apply-mapping ::apply-mapping, :as query}]
+;;;;;;;;;;;;;;;;;;;;;;;;
+;; rule substitution
+;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defrecord ReplacementCtx [var->replacement replacement-fn])
+
+(defn ->replacment-ctx [var->replacement replacement-fn]
```
[the remark this answers is not in the document]
Comment (FiV0) [reply core/src/main/clojure/core2/core/datalog.clj]: Just a habit for `defrecord`'s

File: core/src/main/clojure/core2/core/datalog.clj [hunk 7]
```diff
@@ -309,9 +336,14 @@
 
       :semi-join (sj-term-vars ::semi-join)
       :anti-join (sj-term-vars ::anti-join)
+
       :sub-query (let [{:keys [query]} term-arg]
-                   {:provided-vars (->> (map form-vars (:find query))
-                                        (apply set/union))}))))
+                   {:provided-vars (set/union (->> (:find query)
```
[the remark this answers is not in the document]
Comment (FiV0) [reply core/src/main/clojure/core2/core/datalog.clj]: good catch! Updated.

File: src/test/clojure/core2/core/datalog_test.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (FiV0) [reply src/test/clojure/core2/core/datalog_test.clj]: Removed them.

**Pull request merged**